app/src/main/java/com/epfl/drawyourpath/userProfile/cache/UserModelCached.kt

```diff
@@ -0,0 +1,428 @@
+            surname,
+            UserEntity.fromLocalDateToLong(dateOfBirth),
+            distanceGoal,
+            activityTimeGoal,
+            nbOfPathsGoal,
+            UserEntity.fromBitmapToByteArray(profilePhoto)
+        )
+
+        userModel = fromUserDataToUserModel(userEntity, database)
+
+        CompletableFuture.supplyAsync {
+            cache.insert(userEntity)
+        }.thenAccept {
+            setUserId(userId)
+        }
+    }
+
+    /**
+     * set the user to an existing user
+     * @param userId the user id of the new current user
+     */
+    fun setCurrentUser(userId: String) {
+        checkCurrentUser(false)
+        CompletableFuture.supplyAsync {
+            // insert empty user in cache to avoid null user
+            cache.insertIfEmpty(UserEntity(userId))
+        }.thenComposeAsync {
+            database.getUserAccount(userId)
+        }.thenApplyAsync {
+            cache.insert(fromUserModelToUserData(it))
+        }
+        setUserId(userId)
+    }
+
+    /**
+     * helper function for setting the userId
+     * @param userId the user id
+     */
+    private fun setUserId(userId: String) {
+        currentUserID = userId
+        _currentUserID.value = userId
+    }
+
+    /**
+     * get the current user (read-only).
+     *
+     * Its usage is as follow : `getUser().observe(viewLifecycleOwner) { viewText.text = it.username }`
+     * @return the [LiveData] of [UserEntity]
+     */
+    fun getUser(): LiveData<UserEntity> {
+        checkCurrentUser()
+        return user
+    }
+
+    /**
+     * get userModel
+     * @return the userModel
+     */
+    fun getUserModel(): UserModel {
+        return userModel
+    }
+
+    /**
+     * set the database used for online storing (mainly used for tests)
+     * @param database the database
+     */
+    fun setDatabase(database: Database) {
+        this.database = database
+        // if database is mock then set user to a test user
+        if (database is MockDataBase) {
+            setCurrentUser(database.userIdTest)
+        }
+    }
+
+    /**
+     * This function will return a future with a boolean to know if the username is available in the database
+     * (i.e the userName proposed is not already associated to another user profile)
+     * @param userName userName that the user want to use for his user profile
+     * @return the future that indicate if the username is available
+     */
+    fun isUsernameAvailable(userName: String): CompletableFuture<Boolean> {
+        return database.isUsernameAvailable(userName)
+    }
+
+    /**
+     * Use this function to modify the username(the username will be modify only if it is available on the database)
+     * @param username that we want to set
+     */
+    fun setUsername(username: String): CompletableFuture<Boolean> {
+        checkCurrentUser()
+        return CompletableFuture.supplyAsync {
+            database.setUsername(username).join()
+        }.thenApplyAsync {

```

Comment:
This function seems incorrect. You should separate the setUserName and upadteUsername(). The Modify Username section is break by this function, since the upadteUsername is never called. You should separate setUsername(that must be used only at the profile creation, so it doesn't need to be present in the cache I think ) and upadateUsername(that is used to modify the username in Modify username section ).